File: BlockchainMarketCap/wwwroot/ts/models/ApiResult.ts
/**
 * Enum representing different types of API errors
 */
export enum ApiErrorType {
	/** General error */
	General = "general",

	/** Network error (connectivity issues) */
	Network = "network",

	/** API rate limit exceeded */
	RateLimit = "rateLimit",

	/** Authentication error */
	Authentication = "authentication",

	/** Resource not found */
	NotFound = "notFound",

	/** Server error */
	Server = "server",

	/** Request timeout */
	Timeout = "timeout",
}

/**
 * Interface representing the result of an API operation
 */
export interface ApiResult<T> {
	/** The data returned by the API (null if there was an error) */
	data: T | null;

	/** Indicates whether the API request was successful */
	isSuccess: boolean;

	/** Error message if there was an error, otherwise null */
	errorMessage: string | null;

	/** Error type for categorizing different errors */
	errorType: ApiErrorType | null;
}

/**
 * Creates a successful API result with data
 */
export function createSuccessResult<T>(data: T): ApiResult<T> {
	return {
		data,
		isSuccess: true,
		errorMessage: null,
		errorType: null,
	};
}

/**
 * Creates a failure API result with an error message
 */
export function createFailureResult<T>(
	errorMessage: string,
	errorType: ApiErrorType = ApiErrorType.General,
): ApiResult<T> {
	return {
		data: null,
		isSuccess: false,
		errorMessage,
		errorType,
	};
}

/**
 * Interface for error with response and request properties
 */
interface ApiError {
	response?: {
		status: number;
	};
	request?: unknown;
	code?: string;
	message?: string;
}

/**
 * Creates a failure API result from an Error object
 */
export function createFailureResultFromError<T>(
	error: ApiError,
	defaultMessage = "An unexpected error occurred",
): ApiResult<T> {
	let errorType = ApiErrorType.General;
	let errorMessage = defaultMessage;

	if (error.response) {
		// Server responded with error
		const status = error.response.status;

		if (status === 401 || status === 403) {
			errorType = ApiErrorType.Authentication;
			errorMessage = "Authentication error";
		} else if (status === 404) {
			errorType = ApiErrorType.NotFound;
			errorMessage = "Resource not found";
		} else if (status === 429) {
			errorType = ApiErrorType.RateLimit;
			errorMessage = "API rate limit exceeded";
		} else if (status >= 500) {
			errorType = ApiErrorType.Server;
			errorMessage = "Server error occurred";
		}
	} else if (error.request) {
		// Request made but no response received
		if (error.code === "ECONNABORTED") {
			errorType = ApiErrorType.Timeout;
			errorMessage = "Request timed out";
		} else {
			errorType = ApiErrorType.Network;
			errorMessage = "Network error";
		}
	}

	return createFailureResult<T>(errorMessage, errorType);
}

File: BlockchainMarketCap/Models/ApiResult.cs
using System;

namespace BlockchainMarketCap.Models
{
    /// <summary>
    /// Generic wrapper for API responses that includes success/failure status and error information
    /// </summary>
    /// <typeparam name="T">The type of data returned by the API</typeparam>
    public class ApiResult<T>
    {
        /// <summary>
        /// The data returned by the API (null if there was an error)
        /// </summary>
        public T Data { get; set; }

        /// <summary>
        /// Indicates whether the API request was successful
        /// </summary>
        public bool IsSuccess { get; set; }

        /// <summary>
        /// Error message if there was an error, otherwise null
        /// </summary>
        public string ErrorMessage { get; set; }

        /// <summary>
        /// Error type for categorizing different errors
        /// </summary>
        public ApiErrorType ErrorType { get; set; }

        /// <summary>
        /// Creates a successful result with data
        /// </summary>
        public static ApiResult<T> Success(T data)
        {
            return new ApiResult<T>
            {
                Data = data,
                IsSuccess = true
            };
        }

        /// <summary>
        /// Creates a failure result with an error message
        /// </summary>
        public static ApiResult<T> Failure(string errorMessage, ApiErrorType errorType = ApiErrorType.General)
        {
            return new ApiResult<T>
            {
                IsSuccess = false,
                ErrorMessage = errorMessage,
                ErrorType = errorType
            };
        }

        /// <summary>
        /// Creates a failure result from an exception
        /// </summary>
        public static ApiResult<T> Failure(Exception ex, ApiErrorType errorType = ApiErrorType.General)
        {
            return Failure(ex.Message, errorType);
        }
    }

    /// <summary>
    /// Enum representing different types of API errors
    /// </summary>
    public enum ApiErrorType
    {
        /// <summary>
        /// General error
        /// </summary>
        General,

        /// <summary>
        /// Network error (connectivity issues)
        /// </summary>
        Network,

        /// <summary>
        /// API rate limit exceeded
        /// </summary>
        RateLimit,

        /// <summary>
        /// Authentication error
        /// </summary>
        Authentication,

        /// <summary>
        /// Resource not found
        /// </summary>
        NotFound,

        /// <summary>
        /// Server error
        /// </summary>
        Server,

        /// <summary>
        /// Request timeout
        /// </summary>
        Timeout
    }
}